### PR TITLE
Checkbox magic

### PR DESCRIPTION
**Pull request by agnetedjupvik**
This uses markdown-it-checkbox to support checkboxes for each subtask in a lesson, as adressed in issue [#218 in codeclub-viewer](https://github.com/kodeklubben/codeclub-viewer/issues/218).

The checkboxes here are dumb, and they only have a 1-indexed ID given by markdown-it-checkbox. They can become unique identificators by combining them with the task name (or path). 

Merging this now, while checkboxes aren't in use in the oppgaver repo yet, will change the appearance of regular list items into actual lists, not fake checkboxes. Checkboxes can then be added into oppgaver by using `[ ] ` instead of `+`.

### CODE REVIEW

**Comment by arve0** [comment]
Thanks!

Do you have a build published in somewhere? If not, do not stress it, I'll try it local later today.

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "jquery": "^2.1.3",
     "js-cookie": "^2.0.4",
     "json5": "^0.4.0",
+    "markdown-it-checkbox": "^1.1.0",
```
Comment (arve0) [inline package.json]: Conventions in this repo are "build-deps" => development dependencies and "Client-deps" => dependencies.

Comment (agnetedjupvik) [inline package.json]: Oooh, so I should save-dev on this one. Should have spotted that. Thanks!

**Comment by agnetedjupvik** [comment]
If we can merge this before the weekend, that would be great :)
I've gotten started on a [branch of codeclub-viewer](https://github.com/agnetedjupvik/codeclub-viewer/tree/checkboxState) which introduces some rudimentary actions and reducers which will attempt to add state to these subtasks. I think the work with this would be soo much easier if we could test with these checkboxes, and I want to do all of that next week.

File: styles/lesson.less [hunk 2]
```diff
@@ -289,6 +277,11 @@ input[type="text"] {
   margin-top: 10px;
 }
 
+label {
+  font-weight: normal;
+  margin-left: 6px;
```
Comment (NorwegianKiwi) [inline styles/lesson.less]: Synes ikke disse checkboxene er så veldig fine. Hva om vi heller gjør noe slikt?
```
input[type="checkbox"] {
  opacity: 0;
  filter: alpha(opacity=0);
}
input[type="checkbox"] + label {
  position: relative;
  margin-left: 10px;
  font-weight: normal;
}
input[type="checkbox"] + label:before {
  content: '';
  display: inline-block;
  visibility: visible;
  left: 0;
  width: 1em;
  height: 1em;
  margin: 0 0.8em -0.1em -1em;
  line-height: 0.6;
  text-align: center;
  /* Put checkbox styles here: */
  border: 3px solid #ABDBEA;
}
input[type="checkbox"]:checked + label:before {
  content: '✔';
  line-height: 0.8em;
  text-align: center;
}
```
(fant dette på nettet, og tweaket littegrann, så kan sikkert gjøres bedre, men nå ligner det ihvertfall litt mer på listevisningen)

**Comment by NorwegianKiwi** [comment]
En ting som er litt dumt, er at det ser ut som om markdown-it-checkbox (tror det er denne) ikke parser innholdet helt riktig. For at alt skal komme inni `<label>`, så må det være på samme linje som `[ ]`...

**Comment by arve0** [comment]
Hi! As @NorwegianKiwi says, the style is a step back:

skjermbilde 2017-04-26 kl 21 17 28

What about another solution: I see you plan to persistent store the state. For that you need to implement some click event handler. Then, the same UI be achieved by using similar styling as @NorwegianKiwi suggest.

**Comment by arve0** [comment]
https://jsbin.com/tasubec/edit?html,css,output

**Comment by agnetedjupvik** [comment]
So now I have a .checked class, and I'm thinking that I'll use that for managing the subtask state in codeclub-viewer. 

Why Travis fails and some links appear broken, though, is a mystery to me... I see that we've had problems with it in [earlier commits](https://github.com/arve0/codeclub_lesson_builder/pull/230/commits/00fde82929d1deb2a6762ee65cf7dc04d9e055b5) and it has been solved in later commits for no apparent reason.

**Comment by arve0** [comment]
> Why Travis fails and some links appear broken, though, is a mystery to me...

We use PhantomJS to generate PDFs. PhantomJS crashes sometimes due to a memory leak. The crash is not resolved gracefully. With [headless Chrome](https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md) in the making, hopefully the PDF generation can be done better in near future.

**Comment by arve0** [comment]
Doesn't look the plugin works well will list points which are several lines.

**code**
```md
[ ] Prøv å klikk på de blå klossene midt på skjermen. For eksempel,
  om du klikker på
```

skjermbilde 2017-05-03 kl 22 22 05

As you can see in the picture, only the first line of the check point are contained in the `<label>`.

**Comment by agnetedjupvik** [comment]
Yup, noticed it. I made a [fork](https://github.com/agnetedjupvik/markdown-it-checkbox) of the plugin and will try to see what I can do to fix it in the next couple of days. Don't hesitate to bring in any ideas for that :)

**Comment by arve0** [comment]
Also check the other checkbox plugins, there are two others.

**Comment by agnetedjupvik** [comment]
We found a [plugin](https://github.com/linsir/markdown-it-task-checkbox) which adds all text into labels in order for it to sound nicer using text-to-speech tools. However, we found out that it wrapped everything in so many other tags that it really doesn't give any value to the site at all. So, I'll add the clickable label thing as an issue in my fork, which _can_ be solved later on, but is not required for this to be merged.

**Comment by arve0** [comment]
Still issue with list points over several lines, list points should be contained in a parent node.

For example
```md
[ ] List point
  over several lines
```

should become something like
```html
<li>
  <input type=checkbox>
  <p>list point</p>
  <p>over several lines</p>
</li>
```

but currently renders to
```html
<p><input type=checkbox>list point</p>
<p>over several lines</p>
```

Which gives

skjermbilde 2017-05-09 kl 20 20 17

instead of

skjermbilde 2017-05-09 kl 20 19 57

This makes it hard to distinct "just text" vs "task".

**Comment by NorwegianKiwi** [comment]
Tip: Perhaps correct alignment here as was done for https://github.com/kodeklubben/codeclub-viewer/pull/267 ?

File: styles/lesson.less [hunk 3]
```diff
@@ -289,6 +295,35 @@ input[type="text"] {
   margin-top: 10px;
 }
 
+input[type="checkbox"] {
+  display: none;
+}
+
+input[type="checkbox"] + label {
```
Comment (arve0) [inline styles/lesson.less]: Curious, what does `+ label` and `+ label::before` do?

Comment (arve0) [inline styles/lesson.less]: `<label>` which is sibling to `<input type="checkbox">`.

https://stackoverflow.com/questions/1139763/what-does-the-plus-sign-css-selector-mean

**Comment by arve0** [comment]
Thanks :octocat:

**Pull request merged**